Auto-detect the right conversion

youri-k opened the pull request:
fixes #251 
fixes #177

youri-k commented:
I think you can already review the general structure. I noticed that KNOSSOS cubes were missing in the original issue description, so I'm currently adding support for those.
I'm especially interested in how you like the argument parsing mechanism. I just read that argument parser could also parse known args only, which would could lead to a clearer architecture.

philippotto reviewed (commented):
Thanks for working on this! The current version looks promising!

On a high-level perspective I have some remarks:
- A very common case for the input datasets are multiple layers. For example, I could pass `.../dataset_X` which contains `.../dataset_X/layer_1` and `.../dataset_X/layer_2`. It would be very cool, if this was supported out of the box. (I assume that was also meant with "recursively check the folders" in the original issue by @normanrz). In that scenario, the layer names would be detected automatically.
- overall, we want to be quite robust regarding different input formats (e.g., if the input is wrapped in another folder, it should still be recognized correctly)
- tiled image stacks are not supported, yet, right? these are usually expected to be in a `z/y/x.ext` hierarchy, which would need to be detected so that the tiled cubing code-path can be invoked.
- implementation-wise it would make sense to have a `WkwConverter` parent class with several subclasses, such as: `NiftiConverter`, `ImageStackConverter`, `TiledImageStackConverter` etc. These converter classes could implement an `acceptsInput(input_path: str)` method. then, the main-code could simply ask the converter-classes whether they are able to convert the input. that would be cleaner than deciding from the outside which conversion should be picked. not sure, whether the converter classes should know about the "layers" concept.
- The use cases "nested input folder with multiple layers" and "nested/tiled images in z/y/x" might interfere with each other. We should think this through before implementing it.
- tests would be cool :)

A bit unrelated, but still interesting: should we also support archives as inputs? the cuber could unzip/untar it to /tmp and then convert it. I imagine this to be quite useful for webKnossos' auto-convert feature.

Let me know if you want to discuss this in person!

normanrz commented:
I agree with @philippotto that we need tests for this. There are many different ways how people store their image files. With the tests we can make sure that we support them all and don't break them, when we implement a new pattern.

I also think that the OO approach that @philippotto could help to solve the complexity that this will have. 

I don't think tiled images need to be support right now. Also, I think layers can be supported selectively (e.g. for knossos datasets or single tiff files that have them).

youri-k commented:
@philippotto could you have a look at the image stack converter? This should be ready for review except the multi layer case.

philippotto commented on wkcuber/tile_cubing.py:
```diff
@@ -291,6 +289,10 @@ def tile_cubing(
     )
+    if args is None or not hasattr(args, "dtype") or args.dtype is None:
+        dtype = image_reader.read_dtype(arbitrary_file)
+        if args is not None:
```
If args is None, the dtype isn't passed anywhere if I'm seeing this correctly.

youri-k replied:
Yes, you're correct. :slightly_smiling_face:

philippotto commented on wkcuber/image_readers.py:
```diff
@@ -28,6 +28,9 @@ def read_z_slices_per_file(
         return 1
 
+    def read_dtype(self, file_name: str) -> str:  # pylint: disable=unused-argument
+        return "uint8"
```
why does this default to uint8?

youri-k replied:
This was my workaround because I didn't know how to read the dtype from other libraries. I added a comment that this needs to be done.

philippotto commented on wkcuber/convert_nifti.py:
```diff
@@ -13,47 +13,24 @@
     parser.add_argument(
         "--is_segmentation_layer",
-        "-s",
+        "-sl",
```
does it interfere with the scale flag?

youri-k replied:
Yes

philippotto commented on wkcuber/cubing.py:
```diff
@@ -95,17 +96,25 @@ def create_parser() -> ArgumentParser:
+    input_path = Path(source_path)
+
+    if input_path.is_dir():
+        joined_path = path.join(source_path, "**")
```
the old code only had one `*`. is this intended for cubing nested structures? what happens if I pass a folder which contains two layers? within this pattern matching logic, the layer name extraction would need to happen. otherwise, all input files (of both layers) are thrown together, arent't they? :thinking: or is the `converter` class responsible for dissecting this? then, I'm not sure whether `**` is necessary here..

youri-k replied:
Yes, you're correct again. Initially, I wanted to add the layer config into the cubing model, but now it is handled by the `converter` class, so I can roll back to the old version.

philippotto commented on wkcuber/cubing.py:
```diff
@@ -251,17 +259,24 @@ def cubing(
             file_len=args.wkw_file_len,
         ),
     )
+    # TODO I think this uses a different layer category guessing method
```
todo

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+
+        self.args = args
+
+        # TODO we can detect the layer names correctly, but we currently do not split the source files accordingly
```
todo

philippotto reviewed (commented):
Cool stuff! Looks already very good. I'm not completely done with the `converter.py` review, but already left some comments.

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        for layer, mags in layers.items():
+            layer_names_and_mags[layer] = natsorted(list(mags))[0]
+
+        return dataset_names.pop(), layer_names_and_mags
```
```suggestion
        return dataset_names[0], layer_names_and_mags
```

pop mutates the list, which I would avoid as a habit when possible (even though, it doesn't make a difference here)

youri-k replied:
dataset_names is a `set`, so it is not subscriptable and I check that the set only contains one element. Therefore, I would rather call `pop` than converting the set to a list and then taking the first element of that list.

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+converter_manager = ConverterManager()
+
+if __name__ == "__main__":
+    args = create_parser().parse_args()
+
+    fitting_converters = list(
+        filter(
+            lambda c: c.accepts_input(args.source_path),
+            converter_manager.converter,
+        )
+    )
+
+    if len(fitting_converters) == 0:
+        print(
+            "No converter found. Please specify which converter you want to use or check the source path."
+        )
+        exit(1)
+    elif len(fitting_converters) > 1:
+        print(
+            "Multiple converters found. Check if your source path contains multiple datasets."
+        )
+        exit(1)
+
+    converter = fitting_converters[0]
+    print("Choosing the", converter.__class__.__name__)
+
+    converter.convert_input(args)
```
I would put this in a `main` function and call that here. This avoids "polluting" the global state.

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        ]
+
+
+converter_manager = ConverterManager()
```
This should probably only happen in `__main__` / `main()`

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+if __name__ == "__main__":
+    args = create_parser().parse_args()
+
+    fitting_converters = list(
```
nit: `matching_converters` sounds better to me

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+
+    if len(fitting_converters) == 0:
+        print(
+            "No converter found. Please specify which converter you want to use or check the source path."
```
How would one specify the converter?

youri-k replied:
This was an initial idea of me that you could overwrite the converter checking. However, this should probably be removed or do you think there is a use case for that?

philippotto replied:
I think, there is no strong need for this, since we also have the other high-level entry points for conversion. In the long run, it might make sense to unify all of this, so there is simply `python -m cuber convert ...` with optional `--source_type tiff`. But that's not important for now imo.

philippotto reviewed (changes requested):
Awesome stuff! The tests look promising :rocket:  Left some feedback and questions :)

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+    if name not in dictionary:
+        dictionary[name] = {value}
+    else:
+        dictionary[name].add(value)
```
Ah, thanks for the suggestion.

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+def get_source_files(
+    input_path: str, extensions: Iterable[str], allows_single_file_input: bool
+) -> List[str]:
+    source_path = Path(input_path)
```
Went with the second suggestion because the utils method expects a string

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        return len(source_files) > 0
+
+    def convert_input(self, args: Namespace) -> None:
+        print("Already a WKW dataset. No conversion necessary...")
```
This is a more general question for me: Do we want to have so much console output? I created these `print` calls as validation for me, but maybe they are interesting for the user as well.

philippotto replied:
I think, high-level logging which shows which "insights" the cuber has gathered and which actions it intends to make is useful. I think one needs to decide case-by-case if something is useful. Usually, it's best to validate this by running the cuber and inspecting whether the output seems useful (instead of going through the logging code).

If something is too detailed, one can also switch from `logger.info` to `logger.debug`. Then, that information is only printed when `--verbose` is used :)

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        # Path structure for knossos is .../(dataset_name)/(layer_name)/(mag)folder/x0000/y0000/z0000/filename.raw
+        layers: Dict[str, set] = dict()
+        dataset_names = set()
+        for f in self.source_files:
```
This is an interesting question, yes it does iterate over all files. Since I already extracted all source files, I didn't even think about other traversal methods.

philippotto replied:
I think, the code would be easier to read when an actual tree traversal is done, but let's discuss this in person.

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+
+        layer_names_and_mags = dict()
+
+        for layer, mags in layers.items():
```
> `add_or_update_if_present(layers, "", split_path[0])`
`layers` is dict from layer path to a set of existing mags. Every time we discover a new mag it is inserted into that set, but for the first mag the set needs to be created. That's why I created the helper method, but I see that the method name might be confusing, so I'm happy if you have a suggestion for a better name.

We end up with a set of all mags, which is then converted to a list to be sorted and we extract the "smallest" mag to use the best available quality for the data conversion.

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        print("Converting image stack")
+
+        # add missing config attributes with defaults
+        put_default_if_not_present(args, "target_mag", 1)
+        put_default_if_not_present(args, "wkw_file_len", 32)
+        put_default_if_not_present(args, "interpolation_mode", "default")
+        put_default_if_not_present(args, "start_z", 0)
+        put_default_if_not_present(args, "jobs", cpu_count())
+        put_default_if_not_present(args, "distribution_strategy", "multiprocessing")
+        put_default_if_not_present(args, "job_resources", None)
+        put_default_if_not_present(args, "pad", False)
+        put_default_if_not_present(args, "max_mag", 32)
+        put_default_if_not_present(args, "no_compress", False)
+        put_default_if_not_present(args, "isotropic", False)
+        put_default_if_not_present(args, "verbose", True)
```
Since they are only defined inside the argument parser, I am not sure how to handle this in a more DRY fashion. The first idea that comes to my mind, would be to set the default of every value to `None` in the argument parser and have a similar validator check this after argument parsing. Then we could have a unified default args dict inside each class. Do you think this is a good idea?

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+
+        _, ext = path.splitext(source_files[0])
+
+        assert all(
+            map(lambda p: path.splitext(p)[1] == ext, source_files)
+        ), "Not all image files are of the same type"
```
There was no similar assertion to that previously, but I was anxious to just allow this, since this order of processing depends on the file names, so there could be some inconsistencies. However, if you think this is not a problem, I'll gladly remove that check again.

philippotto replied:
> There was no similar assertion to that previously,

Ok, then let's just keep it :)

youri-k commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        paths = set()
+        for f in self.source_files:
+            p = path.dirname(f)
+            if p != "":
```
p is `""` when the user has only a single input file which is located in the current working directory.

philippotto commented on wkcuber/image_readers.py:
```diff
@@ -28,6 +28,9 @@ def read_z_slices_per_file(
         return 1
 
+    def read_dtype(self, file_name: str) -> str:  # pylint: disable=unused-argument
+        return "uint8"  # TODO: Implement dtype method for all readers
```
todo

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        dictionary[name] = default
+
+
+def add_or_update_if_present(dictionary: dict, name: str, value: Any) -> None:
```
`add_to_set_in_dictionary`

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+
+    def detect_dataset_and_layer_paths_with_mag(self) -> Tuple[str, Dict[str, str]]:
+        # Path structure for knossos is .../(dataset_name)/(layer_name)/(mag)folder/x0000/y0000/z0000/filename.raw
+        layers: Dict[str, set] = dict()
```
`layer_path_to_mag_set`

philippotto commented on tests/test_auto_detection.py:
```diff
@@ -0,0 +1,184 @@
+
+    # test if too short absolute path gets detected
+    converter.source_files = [
+        "/y0000/z0000/test_mag1_x0000_y0000_z0000.raw",
```
maybe do all test cases with these three scenarios:
- without leading slash
- with leading "../"
- with leading "/"

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,283 @@
+        convert_image_stack(args)
+
+    def detect_dataset_and_layer_names(self) -> Tuple[str, List[str]]:
+        paths = set()
```
`parent_paths_of_images`

youri-k commented:
@philippotto I think this is ready for another round of review :slightly_smiling_face: I hope I addressed all of your suggestions and I'm happy to hear more feedback.

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,461 @@
+    def convert_input(self, args: Namespace) -> None:
+        pass
+
+    # returns [prefix, traversal_depth]
```
seems outdated? can probably be removed.
```suggestion
```

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,461 @@
+                lambda p: len(p.split(sep)) == traversal_depth,
+                self.source_files,
+            )
+        )
```
please add an assertion message which explains what went wrong (`assert ..., "msg"`)

philippotto commented on wkcuber/converter.py:
```diff
@@ -0,0 +1,461 @@
+                lambda p: len(p.split(sep)) == traversal_depth,
+                self.source_files,
+            )
+        )
```
please add an assertion message

philippotto reviewed (approved):
Great work! Thanks for pulling this through the finish line. I left some cosmetic feedback, but nothing major.

Could you create a follow-up issue that the converter classes should also be used when not using `__main__` (but for example `wkcuber.cubing`). That way, we can further refactor the argparse code, I hope.

Also, please update the readme in the following section, so that it explains the auto-convert behavior.

> wkcuber: Convert image stacks to fully ready WKW datasets (includes downsampling, compressing and metadata generation)

youri-k commented:
@philippotto Can you have one last look? I hope this is good to go now :slightly_smiling_face:

philippotto commented on README.md:
```diff
@@ -10,10 +10,11 @@ The tools are modular components to allow easy integration into existing pipelin
-* `wkcuber.cubing`: Convert image stacks (e.g., `tiff`, `jpg`, `png`, `dm3`) to WKW cubes
+* `wkcuber.cubing`: Convert image stacks (e.g., `tiff`, `jpg`, `png`, `dm3`, `dm4`) to WKW cubes
 * `wkcuber.tile_cubing`: Convert tiled image stacks (e.g. in `z/y/x.ext` folder structure) to WKW cubes
+* `wkcuber.convert_image_stack_to_wkw`: Convert image stacks to fully ready WKW datasets (includes downsampling, compressing and metadata generation)
```
I'd move that directly below the `wkcuber` part :)

philippotto commented on wkcuber/__main__.py:
```diff
@@ -36,52 +57,49 @@ def create_parser() -> ArgumentParser:
+            downsample_mags_anisotropic(
+                args.target_path,
+                layer_name,
+                mags[0],
```
shouldn't it be this?
```suggestion
                mags[-1],
```

Otherwise, mag 1 would overwrite the existing mag 2 if both exist? Also: Is the mags list natsorted? This would be necessary for this to work.

philippotto replied:
Ah, I missed line 73. Alrighty then. I'd prefer to not use `reverse=True` and instead access `[-1]`. But you can leave it as is.

However: Wouldn't the sorting need to be natsorted? For example, sorting `[1, 8, 16]` wouldn't work otherwise, I think.

youri-k replied:
As far as I tested it, it works. My guess would be that the `Mag` objects that we're using have an internal correct sorting method (or less-equal method). I went with your suggestion and accessed `[-1]`

philippotto replied:
> My guess would be that the Mag objects that we're using have an internal correct sorting method (or less-equal method).

Ah, you are right! I thought of them as strings, but that was wrong :) The class indeed implements `__le__` and co :thinking:

philippotto reviewed (approved):
Awesome work! See my last few comments, but after that, feel free to merge :)
